internal/extension_inspector.go

```diff
@@ -0,0 +1,119 @@
+package internal
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/paketo-buildpacks/packit/v2/cargo"
+)
+
+type ExtensionInspector struct{}
+
+func NewExtensionInspector() ExtensionInspector {
+	return ExtensionInspector{}
+}
+
+type ExtensionMetadata struct {
+	Config cargo.ExtensionConfig
+	SHA256 string
+}
+
+func (i ExtensionInspector) Dependencies(path string) ([]ExtensionMetadata, error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	indicesJSON, err := fetchFromArchive(tar.NewReader(file), "index.json", true)
+	if err != nil {
+		return nil, err
+	}
+
+	var index struct {
+		Manifests []struct {
+			Digest string `json:"digest"`
+		} `json:"manifests"`
+	}
+
+	// There can only be 1 image index
+	err = json.NewDecoder(indicesJSON[0]).Decode(&index)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = file.Seek(0, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	manifests, err := fetchFromArchive(tar.NewReader(file), filepath.Join("blobs", "sha256", strings.TrimPrefix(index.Manifests[0].Digest, "sha256:")), true)
+	if err != nil {
+		return nil, err
+	}
+
+	buildpackageDigest := index.Manifests[0].Digest
+
+	var m struct {
+		Layers []struct {
+			Digest string `json:"digest"`
+		} `json:"layers"`
+	}
+
+	// We only support single manifest images
+	err = json.NewDecoder(manifests[0]).Decode(&m)
+	if err != nil {
+		return nil, err
+	}
+
+	var metadataCollection []ExtensionMetadata
+	for _, layer := range m.Layers {
+		_, err = file.Seek(0, 0)
+		if err != nil {
+			return nil, err
+		}
+
+		layerBlobs, err := fetchFromArchive(tar.NewReader(file), filepath.Join("blobs", "sha256", strings.TrimPrefix(layer.Digest, "sha256:")), true)
+		if err != nil {
+			return nil, err
+		}
+
+		layerGR, err := gzip.NewReader(layerBlobs[0])
+		if err != nil {
+			return nil, fmt.Errorf("failed to read layer blob: %w", err)
+		}
+		defer layerGR.Close()
+
+		// Generally, each layer corresponds to a buildpack.
```

Comment:
I'd update this comment to describe extensions, not buildpacks. Even though the words are the same :)

Reply:
Thanks for taking the time to have a look :)  We discussed using a more generic implementation to avoid duplicate code, I'm also a big fan of this idea. However, with extensions, we are still determining some use cases that buildpacks support on whether extensions should supported them too. E.g. on the code you have commented, the question is, can an extension have multiple extensions, similar to what buildpacks do? It is possible but we don't have an actual use case that this would be necessary or would make sense. So, having a generic implementation when there are several grey areas will end up with a messy/hacky code. Therefore, we ended up keeping things separate till all scenarios around extensions compared to buildpacks have been covered.

Reply:
@robdimsdale renamed the variable and also rebase the branch to main with force push